File: src/vault/template.ts
import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { parseNote, serializeNote } from './frontmatter.js';

export interface TemplateVars {
  title: string;
  date: string;       // YYYY-MM-DD
  [key: string]: string;
}

/**
 * Load a template file from templatesDir.
 * Template name should not include .md extension.
 */
export async function loadTemplate(templatesDir: string, templateName: string): Promise<string> {
  const templatePath = path.join(templatesDir, `${templateName}.md`);

  try {
    return await readFile(templatePath, 'utf-8');
  } catch (err) {
    throw new Error(`Failed to load template "${templateName}": ${(err as Error).message}`);
  }
}

/**
 * Replace template placeholders with variables.
 * Handles:
 * - <% tp.file.title %> → vars.title
 * - <% tp.date.now("YYYY-MM-DD") %> → vars.date
 * - Other <% tp.date.now("...") %> patterns with vars.date
 */
export function scaffoldNote(templateContent: string, vars: TemplateVars): string {
  let result = templateContent;

  // Replace <% tp.file.title %>
  result = result.replace(/<% tp\.file\.title %>/g, vars.title);

  // Replace <% tp.date.now("...") %> with vars.date
  result = result.replace(/<% tp\.date\.now\("[^"]*"\) %>/g, vars.date);

  return result;
}

/**
 * Merge existing note content into template sections.
 * For each ## Section in the template, look for matching content in existingContent.
 * If found, place it in the template. If not, leave placeholder and mark with <!-- TODO -->.
 */
export function mergeContentIntoTemplate(templateContent: string, existingContent: string): string {
  const templateLines = templateContent.split('\n');
  const existingLines = existingContent.split('\n');

  // Extract sections from template
  interface TemplateSection {
    heading: string;
    index: number;
  }
  const sections: TemplateSection[] = [];
  for (let i = 0; i < templateLines.length; i++) {
    if (templateLines[i].startsWith('## ')) {
      sections.push({
        heading: templateLines[i],
        index: i,
      });
    }
  }

  if (sections.length === 0) {
    // No sections in template, just append existing content
    return templateContent + '\n\n' + existingContent;
  }

  // Extract sections from existing content
  interface ExistingSection {
    heading: string;
    content: string;
  }
  const existingSections: Map<string, ExistingSection> = new Map();
  for (let i = 0; i < existingLines.length; i++) {
    if (existingLines[i].startsWith('## ')) {
      const heading = existingLines[i];
      const endIndex = existingLines.findIndex(
        (line, idx) => idx > i && line.startsWith('## ')
      );
      const contentEnd = endIndex === -1 ? existingLines.length : endIndex;
      const content = existingLines.slice(i + 1, contentEnd).join('\n').trim();

      existingSections.set(heading, { heading, content });
    }
  }

  // Build output by iterating through template and replacing sections
  const output: string[] = [];
  let lastSectionEnd = 0;

  for (let i = 0; i < sections.length; i++) {
    const section = sections[i];
    const nextSectionStart = i + 1 < sections.length ? sections[i + 1].index : templateLines.length;

    // Add everything from last section end to this section
    output.push(...templateLines.slice(lastSectionEnd, section.index));

    // Add section heading
    output.push(section.heading);

    // Check if existing content has this section
    const existing = existingSections.get(section.heading);
    if (existing && existing.content) {
      output.push(existing.content);
    } else {
      // Mark with TODO
      output.push('<!-- TODO -->');
    }

    lastSectionEnd = nextSectionStart;
  }

  // Add any remaining template lines
  output.push(...templateLines.slice(lastSectionEnd));

  return output.join('\n').trimEnd();
}

File: internal/vault/template.go
package vault

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// TemplateVars holds substitution values for template scaffolding.
type TemplateVars struct {
	Title string // note title (filename without .md)
	Date  string // YYYY-MM-DD
}

// LoadTemplate reads a template file from templatesDir.
// templateName should not include .md extension.
func LoadTemplate(templatesDir, templateName string) (string, error) {
	templatePath := filepath.Join(templatesDir, templateName+".md")
	data, err := os.ReadFile(templatePath)
	if err != nil {
		return "", fmt.Errorf("failed to load template %q: %w", templateName, err)
	}
	return string(data), nil
}

var (
	reTitleVar = regexp.MustCompile(`<%\s*tp\.file\.title\s*%>`)
	reDateVar  = regexp.MustCompile(`<%\s*tp\.date\.now\("[^"]*"\)\s*%>`)
)

// ScaffoldNote replaces Obsidian Templater placeholders with vars.
// Handles:
//   - <% tp.file.title %> → vars.Title
//   - <% tp.date.now("...") %> → vars.Date
func ScaffoldNote(templateContent string, vars TemplateVars) string {
	result := reTitleVar.ReplaceAllString(templateContent, vars.Title)
	result = reDateVar.ReplaceAllString(result, vars.Date)
	return result
}

// MergeContentIntoTemplate merges existing note content into template sections.
// For each ## Section in the template, looks for matching content in existingContent.
// Missing sections are marked with <!-- TODO -->.
func MergeContentIntoTemplate(templateContent, existingContent string) string {
	templateLines := strings.Split(templateContent, "\n")
	existingLines := strings.Split(existingContent, "\n")

	// Extract ## sections from template
	type section struct {
		heading string
		index   int
	}
	var sections []section
	for i, line := range templateLines {
		if strings.HasPrefix(line, "## ") {
			sections = append(sections, section{heading: line, index: i})
		}
	}

	if len(sections) == 0 {
		// No sections — just append existing content
		return templateContent + "\n\n" + existingContent
	}

	// Extract ## sections from existing content
	existingSections := map[string]string{}
	for i, line := range existingLines {
		if strings.HasPrefix(line, "## ") {
			heading := line
			// Find end of section (next ## or end)
			end := len(existingLines)
			for j := i + 1; j < len(existingLines); j++ {
				if strings.HasPrefix(existingLines[j], "## ") {
					end = j
					break
				}
			}
			sectionContent := strings.TrimSpace(strings.Join(existingLines[i+1:end], "\n"))
			existingSections[heading] = sectionContent
		}
	}

	// Build output
	var output []string
	lastSectionEnd := 0

	for i, sec := range sections {
		nextSectionStart := len(templateLines)
		if i+1 < len(sections) {
			nextSectionStart = sections[i+1].index
		}

		// Add lines before this section
		output = append(output, templateLines[lastSectionEnd:sec.index]...)

		// Add section heading
		output = append(output, sec.heading)

		// Add content or TODO placeholder
		if content, ok := existingSections[sec.heading]; ok && content != "" {
			output = append(output, content)
		} else {
			output = append(output, "<!-- TODO -->")
		}

		lastSectionEnd = nextSectionStart
	}

	// Add remaining template lines
	output = append(output, templateLines[lastSectionEnd:]...)

	return strings.TrimRight(strings.Join(output, "\n"), "\n")
}
